### ISSUE TITLE
Mask misalignment in prediction output

### ISSUE THREAD

**Issue by horsto** [ultralytics/ultralytics#20918]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

I notice an offset in my prediction results. Most pronounced on the y axis: The predicted masks seem to be shifted towards the bottom of the current frame. 
Similar issues have been reported https://github.com/ultralytics/ultralytics/issues/11794 and (older) here https://github.com/ultralytics/ultralytics/issues/2584 . 
In 11794 the issue was marked as solved with an update of ultralytics at the time, but I don't think this was solved. 
I am using augmentation in my training. I tried to use rect=True, thinking that the rectangular input images might cause issues. This however did not solve this issue. 

Any ideas? 

![Image](https://github.com/user-attachments/assets/b40f5b5a-a744-4cbe-b015-60b1adfaf84f)

### Environment

Ultralytics 8.3.127 🚀 Python-3.11.11 torch-2.6.0 CPU (Apple M2 Max)
Setup complete ✅ (12 CPUs, 96.0 GB RAM, 1326.3/1858.2 GB disk)

OS                  macOS-15.5-arm64-arm-64bit
Environment         Darwin
Python              3.11.11
Install             pip
Path                /Users/horst/miniconda3/envs/octron/lib/python3.11/site-packages/ultralytics
RAM                 96.00 GB
Disk                1326.3/1858.2 GB
CPU                 Apple M2 Max
CPU count           12
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 2.1.1>=1.23.0
matplotlib          ✅ 3.10.1>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.6.0>=1.8.0
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

**Comment by UltralyticsAssistant** [comment]
👋 Hello @horsto, thank you for bringing this to our attention and for providing a detailed report 🚀! This is an automated response to help streamline troubleshooting—an Ultralytics engineer will also review your issue soon.

If this is a 🐛 Bug Report, please ensure you’ve included a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug the mask misalignment you’re experiencing. If possible, please provide a simple code snippet and sample data that clearly demonstrate the issue.

If you’re seeking advice on custom training, please share additional details such as dataset image examples and your training logs, and review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Check out the [Docs](https://docs.ultralytics.com/) for more info, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples.

Join the Ultralytics community for support or discussion:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com/) for in-depth topics
- [Subreddit](https://reddit.com/r/Ultralytics) for sharing knowledge

## Upgrade

To ensure the issue isn’t resolved in a more recent release, please upgrade to the latest `ultralytics` package (with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in these up-to-date, verified environments (preinstalled dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/)):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your detailed report!

**Comment by Y-T-G** [comment]
Can you report if this fixes the issue?

https://github.com/ultralytics/ultralytics/issues/18326#issuecomment-2561647951

**Comment by horsto** [comment]
Thanks @Y-T-G , but I don't think this will be solved by simple cropping. There is a systematic shift downwards of masks vs. original image that is apparent in the mask outline itself. The training data / polygons don't have that shift and it is not an artefact of post-processing on my side since the internal visualization tools of ultralytics during prediction show the same. 

![Image](https://github.com/user-attachments/assets/77e2fad4-e434-47a8-bab5-727cd56b6312)

**Comment by horsto** [comment]
very similar to https://github.com/ultralytics/ultralytics/issues/2584#issuecomment-1598025965 
and you can see it in @rob-safi 's masks too https://github.com/ultralytics/ultralytics/issues/18326#issue-2752361965 
(see yellow boxes top vs. bottom - the bottom includes more, and the top less of the object)
<img width="816" alt="Image" src="https://github.com/user-attachments/assets/bcf8f1c8-93c8-4858-83f6-99f8ca2fe5b5" />

**Comment by horsto** [comment]
Same for example @SteFehr posted in https://github.com/ultralytics/ultralytics/issues/2584 

<img width="609" alt="Image" src="https://github.com/user-attachments/assets/95a166aa-c9dd-4521-8dbd-b0bb24bc0d62" />

**Comment by Y-T-G** [comment]
Yes, it seems like a post-processing error. The function I provided casts to `int`. I wanted to see if it helps because it may be due to rounding error.

**Comment by horsto** [comment]
Yeah I tried all kinds of things using xy or xyn (normalized) coordinates and different ways of rounding - but the problem persists. Also, rounding itself would not induce a systematic (and exclusive) y shift I think.

**Comment by Y-T-G** [comment]
What if you use `retina_masks=True` during prediction?

**Comment by horsto** [comment]
Same result

**Comment by Y-T-G** [comment]
The issue doesn't occur with pretrained models, so I can't reproduce it to debug.

**Comment by horsto** [comment]
Ok, but many others seem to have the same issue when training their model. Any other ideas how to debug?

**Comment by Y-T-G** [comment]
What was your training command?

**Comment by Y-T-G** [comment]
Can you try training with Ultralytics 8.3.0?

**Comment by horsto** [comment]
Training is done via 

```python
self.model.train(
                    data=self.config_path.as_posix(), 
                    name='training',
                    project=self.training_path.as_posix(),
                    mode='segment',
                    device=device,
                    optimizer='auto',
                    rect=True,
                    cos_lr=True,
                    mask_ratio=2,
                    overlap_mask=True,
                    fraction=1.0,
                    epochs=epochs,
                    imgsz=imagesz,
                    resume=False,
                    patience=50,
                    plots=True,
                    batch=-1,
                    cache='disk',
                    save=True,
                    save_period=save_period, 
                    exist_ok=True,
                    nms=False, 
                    # Augmentation
                    hsv_v=.25,
                    hsv_s=0.25,
                    hsv_h=0.25,
                    degrees=180,
                    translate=0.1,
                    perspective=0,
                    scale=0.25,
                    shear=2,
                    flipud=.5,
                    fliplr=.5,
                    mosaic=0.25,
                    mixup=0.25,
                    copy_paste=0.25,
                    copy_paste_mode='mixup', 
                    erasing=0.,
                )

```
where 
`imagesz` is either 640 or 1024 


---
prediction is done via 

```python

results = model.track(
                    source=CURRENT_FRAME, 
                    persist=True,
                    task='segment',
                    tracker=tracker_yaml_path.as_posix(),
                    project=save_dir.parent.as_posix(),
                    name=save_dir.name,
                    show=False,
                    #show_boxes=False, # show=False!
                    save=False,
                    verbose=False,
                    imgsz=imgsz,
                    stream_buffer=True,
                    conf=conf_thresh,
                    iou=iou_thresh,
                    device=device, 
                    retina_masks=retina_masks, # original image resolution, not inference resolution
                    save_txt=False,
                    save_conf=False,
                )
```

**Comment by horsto** [comment]
> Can you try training with Ultralytics 8.3.0?

I can give that a go, yes

**Comment by Y-T-G** [comment]
Did you try without adjusting mask ratio?

**Comment by horsto** [comment]
I think it's set to 4 originally and I set it to 2. Should I try with 4?

**Comment by horsto** [comment]
I was hoping to gain higher resolution training by setting it to 2.

**Comment by Y-T-G** [comment]
You can try 4.

**Comment by horsto** [comment]
Small update: I have now, downgraded to `Ultralytics 8.3.0` and used mask ratio 4 (instead of 2 before). 

<img width="725" alt="Image" src="https://github.com/user-attachments/assets/4b3dc599-0ecf-4972-abc0-61b43c7b293f" />
<img width="484" alt="Image" src="https://github.com/user-attachments/assets/5e74461f-2c8c-4535-b87c-e34ff3042369" />

The problem does persist under these conditions. Note that the effect is really subtle: I marked it with boxes in the images. It seems to be a stretching in y mostly.

**Comment by horsto** [comment]
Following @glenn-jocher 's comment in the other thread: 

> @horsto your observation about rect=True/False is insightful - rectangular training can indeed cause alignment issues when aspect ratios differ significantly from 1:1. Try setting rect=False during training if using non-square images, or ensure your inference images match the training aspect ratio.
> 
> For a quick diagnostic, you could check if the offset correlates with your image dimensions not being multiples of the model's stride (typically 32). Could you both share your exact image dimensions and model version? This would help pinpoint if it's a specific resolution or version-related issue.


rect=True/False does not seem to affect the result. 

I have been using the YOLO m and l models for training. 
It seems to be more pronounced in the earlier iterations I tried (with up to date ultralytics and mask ratio = 2, rect=False). 

<img width="308" alt="Image" src="https://github.com/user-attachments/assets/fe9ef4c0-71ba-4092-91f1-505be4ac4f4a" />


The particular image size I am trying to predict on (which matches the training image dimensions) is 2064 × 1544. However, I do seem to be to reproduce this scaling effect also in square images. 

I am lacking intuition of what could be going on there. It seems likely to me that there is some scaling in y (and possibly x) going on.

**Comment by horsto** [comment]
One thing I notice is a  small difference between retina_masks True vs. False

<img width="754" alt="Image" src="https://github.com/user-attachments/assets/c867241d-3a3c-416d-b33a-2dad875546b7" />

The retina_masks=False condition gives jagged edges, but it seems that the shift is less pronounced (could simply be a resolution problem?).

**Comment by Y-T-G** [comment]
Can you try running prediction with `rect=False`?

**Comment by horsto** [comment]
> Can you try running prediction with rect=False?


imgsz 640 trained model: 
<img width="394" alt="Image" src="https://github.com/user-attachments/assets/92a3a9d6-59ff-4b9c-9997-0ec5ad19734b" />

imgsz 1024 trained model: 
<img width="385" alt="Image" src="https://github.com/user-attachments/assets/070c680e-ac42-47ed-b2db-b823e8061c9f" />

Same effect. I tried setting it (during prediction) to False and True -> identical result.

**Comment by Y-T-G** [comment]
Maybe it's related to upscaling. The masks are 1/4th the size of input image, so they get upscaled. Maybe the upscaling algorithm shifts it.

Otherwise the problem might be deeper during training.

**Comment by horsto** [comment]
So I did try using the xyn (normalized xy) coordinates and played with the scaling, but nothing logical popped out to me. 

What about before that, i.e. when mask size / input image size / model stride size collide somehow?

**Comment by Y-T-G** [comment]
The original output are the masks at `result.masks.data`. The `xy` and `xyn` are just derived from that.

**Comment by Y-T-G** [comment]
Can you try setting `align_corners=True` here:

https://github.com/ultralytics/ultralytics/blob/487e27639595047cff8775dab5e2ff268d8647c4/ultralytics/utils/ops.py#L763

https://github.com/ultralytics/ultralytics/blob/487e27639595047cff8775dab5e2ff268d8647c4/ultralytics/utils/ops.py#L717

**Comment by horsto** [comment]
Tried this right now, no difference.

**Comment by horsto** [comment]
We have consolidated more work on this in a PR https://github.com/ultralytics/ultralytics/pull/20957 that improves the alignment issue quite a bit.

**Comment by glenn-jocher** [comment]
Thanks for the comprehensive debugging work and for creating PR #20957 to address the mask alignment issue! It's great to see you've identified improvements that help with the mask shifting problem. The PR approach is the best way to contribute fixes back to the community - I appreciate you taking the initiative to work on a solution rather than just reporting the issue.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #20957](https://github.com/ultralytics/ultralytics/pull/20957): **ultralytics 8.3.152 Optimize padding for Segment mask processing**. The update improves mask scaling accuracy by refining how padding and cropping are handled in image processing functions. By updating `scale_image` and `scale_masks` to use `round()` for boundary calculations, the changes aim to deliver more precise mask alignment and reduce off-by-one errors—especially beneficial for segmentation edge cases.

Big thanks to @horsto for authoring this PR, with valuable contributions from @glenn-jocher, @Y-T-G, @Laughing-q, and @baptist.

To test if the fix resolves your issue, you can upgrade to the latest version with:
```
pip install -U ultralytics>=8.3.152
```
Alternatively, try the latest code directly from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
You may also wait for the next official PyPI release if you prefer.

Please let us know if these changes resolve the issue or if any problems persist. Thanks again for reporting this, and any further feedback is always welcome! 🙏
